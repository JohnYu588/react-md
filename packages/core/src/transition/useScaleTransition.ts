import type {
  CSSTransitionClassNames,
  CSSTransitionClassNamesObject,
  CSSTransitionHookReturnValue,
  PreconfiguredCSSTransitionOptions,
  TransitionTimeoutObject,
} from "./types";
import { useCSSTransition } from "./useCSSTransition";

/**
 * The default {@link CSSTransitionClassNames} for a horizontal scale
 * transition.
 *
 * @remarks
 * \@since 2.0.0
 * \@since 6.0.0 The class names were updated to be prefixed with
 * `rmd-scale-transition`
 */
export const SCALE_CLASSNAMES: Readonly<CSSTransitionClassNamesObject> = {
  appear: "rmd-scale-transition--enter",
  appearActive: "rmd-scale-transition--enter-active",
  enter: "rmd-scale-transition--enter",
  enterActive: "rmd-scale-transition--enter-active",
  enterDone: "",
  exit: "rmd-scale-transition--exit",
  exitActive: "rmd-scale-transition--exit-active",
};

/**
 * The default {@link CSSTransitionClassNames} for a vertical scale transition.
 *
 * @remarks
 * \@since 2.0.0
 * \@since 6.0.0 The class names were updated to be prefixed with
 * `rmd-scale-y-transition` and merged with the {@link SCALE_CLASSNAMES}
 */
export const SCALE_Y_CLASSNAMES: Readonly<CSSTransitionClassNamesObject> = {
  appear: "rmd-scale-transition--enter rmd-scale-transition--y-enter",
  appearActive:
    "rmd-scale-transition--enter-active rmd-scale-transition--y-enter-active",
  enter: "rmd-scale-transition--enter rmd-scale-transition--y-enter",
  enterActive:
    "rmd-scale-transition--enter-active rmd-scale-transition--y-enter-active",
  enterDone: "",
  exit: "rmd-scale-transition--exit rmd-scale-transition--y-exit",
  exitActive:
    "rmd-scale-transition--exit-active rmd-scale-transition--y-exit-active",
};

/**
 * The default {@link TransitionTimeout} to use for horizontal and vertical
 * scale transitions.
 *
 * @remarks \@since 2.0.0
 */
export const SCALE_TIMEOUT: Readonly<TransitionTimeoutObject> = {
  enter: 200,
  exit: 150,
};

/**
 * @typeParam E - The HTMLElement type used or the ref required for the
 * transition.
 * @remarks \@since 4.0.0
 */
export interface ScaleTransitionHookOptions<E extends HTMLElement>
  extends PreconfiguredCSSTransitionOptions<E> {
  /**
   * Boolean if the scale transition should be vertical instead of horizontal.
   * This really only changes the default value for the {@link classNames}.
   *
   * @defaultValue `false`
   */
  vertical?: boolean;

  /**
   * @see {@link PreconfiguredCSSTransitionOptions.temporary}
   * @defaultValue `true`
   */
  temporary?: boolean;

  /**
   * @see {@link vertical}
   * @see {@link SCALE_CLASSNAMES}
   * @see {@link SCALE_Y_CLASSNAMES}
   * @defaultValue `vertical ? SCALE_Y_CLASSNAMES : SCALE_CLASSNAMES`
   */
  classNames?: CSSTransitionClassNames;
}

/**
 * Implements a scale transition that should generally be used for temporary
 * elements that are positioned via `position: absolute` or `position: fixed`.
 *
 * @example
 * Dropdown Menu Example
 * ```tsx
 * import { ReactElement, useRef, useState } from "react";
 * import { Button, useFixedPositioning, useScaleTransition } from "@react-md/core";
 *
 * function Example(): ReactElement {
 *   const buttonRef = useRef<HTMLButtonElement>(null);
 *   const [transitionIn, setTransitionIn] = useState(false);
 *   const { style, transitionOptions } = useFixedPositioning({
 *     fixedTo: buttonRef,
 *   });
 *   const { elementProps, rendered } = useScaleTransition({
 *     ...transitionOptions,
 *     transitionIn,
 *     vertical: true,
 *   });
 *
 *   return (
 *     <>
 *       <Button ref={buttonRef} onClick={() => setTransitionIn(!transitionIn)}>
 *         Toggle
 *       </Button>
 *       {rendered && (
 *         <div {...elementProps} style={style}>
 *           Some content within a menu
 *         </div>
 *       )}
 *     </>
 *   );
 * }
 * ```
 *
 * @typeParam E - The HTMLElement type used or the ref required for the
 * transition.
 * @remarks \@since 4.0.0
 */
export function useScaleTransition<E extends HTMLElement>(
  options: ScaleTransitionHookOptions<E>
): CSSTransitionHookReturnValue<E> {
  const {
    timeout = SCALE_TIMEOUT,
    vertical = false,
    temporary = true,
    classNames = vertical ? SCALE_Y_CLASSNAMES : SCALE_CLASSNAMES,
    ...transitionOptions
  } = options;

  return useCSSTransition({
    ...transitionOptions,
    timeout,
    temporary,
    classNames,
  });
}
