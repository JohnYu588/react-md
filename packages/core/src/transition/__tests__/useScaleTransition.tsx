import {
  afterEach,
  beforeAll,
  describe,
  expect,
  it,
  jest,
} from "@jest/globals";
import type { ReactElement } from "react";
import { useState } from "react";
import { act, fireEvent, render } from "../../test-utils/index.js";

import type { ScaleTransitionHookOptions } from "../useScaleTransition.js";
import { useScaleTransition } from "../useScaleTransition.js";

interface TestProps
  extends Omit<ScaleTransitionHookOptions<HTMLElement>, "transitionIn"> {
  defaultTransitionIn?: boolean;
}

function Test({
  defaultTransitionIn = false,
  ...options
}: TestProps): ReactElement {
  const [transitionIn, setTransitionIn] = useState(defaultTransitionIn);
  const { elementProps, rendered } = useScaleTransition({
    transitionIn,
    ...options,
  });
  return (
    <>
      <button type="button" onClick={() => setTransitionIn((p) => !p)}>
        Toggle
      </button>
      {rendered && (
        <div {...elementProps} data-testid="element">
          This is some content.
        </div>
      )}
    </>
  );
}

describe("ScaleTransition", () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllTimers();
  });

  it("should default using the horizontal transition", () => {
    const { container, getByRole, getByTestId } = render(<Test />);
    const getElement = (): HTMLElement => getByTestId("element");
    const toggle = getByRole("button");

    expect(getElement).toThrow();
    expect(container).toMatchSnapshot();

    fireEvent.click(toggle);
    expect(getElement).not.toThrow();
    expect(container).toMatchSnapshot();

    act(() => {
      jest.runAllTimers();
    });
    expect(getElement).not.toThrow();
    expect(container).toMatchSnapshot();

    fireEvent.click(toggle);
    expect(getElement).not.toThrow();
    expect(container).toMatchSnapshot();

    act(() => {
      jest.runAllTimers();
    });
    expect(getElement).toThrow();
    expect(container).toMatchSnapshot();
  });

  it("should render correctly for a vertical transition", () => {
    const { container, getByRole, getByTestId } = render(<Test vertical />);
    const getElement = (): HTMLElement => getByTestId("element");
    const toggle = getByRole("button");

    expect(getElement).toThrow();
    expect(container).toMatchSnapshot();

    fireEvent.click(toggle);
    expect(getElement).not.toThrow();
    expect(container).toMatchSnapshot();

    act(() => {
      jest.runAllTimers();
    });
    expect(getElement).not.toThrow();
    expect(container).toMatchSnapshot();

    fireEvent.click(toggle);
    expect(getElement).not.toThrow();
    expect(container).toMatchSnapshot();

    act(() => {
      jest.runAllTimers();
    });
    expect(getElement).toThrow();
    expect(container).toMatchSnapshot();
  });

  it("should merge the className and allow to be rendered by default", () => {
    const { container, getByRole, getByTestId } = render(
      <Test className="custom-class" defaultTransitionIn />
    );

    const getElement = (): HTMLElement => getByTestId("element");
    const toggle = getByRole("button");

    expect(getElement()).toHaveClass("custom-class");
    expect(container).toMatchSnapshot();

    fireEvent.click(toggle);
    expect(getElement()).toHaveClass("custom-class");
    expect(container).toMatchSnapshot();

    act(() => {
      jest.runAllTimers();
    });
    expect(getElement).toThrow();
    expect(container).toMatchSnapshot();

    fireEvent.click(toggle);
    expect(getElement()).toHaveClass("custom-class");
    expect(container).toMatchSnapshot();

    act(() => {
      jest.runAllTimers();
    });
    expect(getElement()).toHaveClass("custom-class");
    expect(container).toMatchSnapshot();
  });

  it("should correctly call the transition callbacks", () => {
    const onEnter = jest.fn();
    const onEntering = jest.fn();
    const onEntered = jest.fn();
    const onExit = jest.fn();
    const onExiting = jest.fn();
    const onExited = jest.fn();
    const props: TestProps = {
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
    };

    const { getByRole } = render(<Test {...props} />);
    const toggle = getByRole("button");

    expect(onEnter).not.toHaveBeenCalled();
    expect(onEntering).not.toHaveBeenCalled();
    expect(onEntered).not.toHaveBeenCalled();
    expect(onExit).not.toHaveBeenCalled();
    expect(onExiting).not.toHaveBeenCalled();
    expect(onExited).not.toHaveBeenCalled();

    fireEvent.click(toggle);
    expect(onEnter).toHaveBeenCalledTimes(1);
    expect(onEntering).toHaveBeenCalledTimes(1);
    expect(onEntered).not.toHaveBeenCalled();
    expect(onExit).not.toHaveBeenCalled();
    expect(onExiting).not.toHaveBeenCalled();
    expect(onExited).not.toHaveBeenCalled();

    act(() => {
      jest.runAllTimers();
    });
    expect(onEnter).toHaveBeenCalledTimes(1);
    expect(onEntering).toHaveBeenCalledTimes(1);
    expect(onEntered).toHaveBeenCalledTimes(1);
    expect(onExit).not.toHaveBeenCalled();
    expect(onExiting).not.toHaveBeenCalled();
    expect(onExited).not.toHaveBeenCalled();

    fireEvent.click(toggle);
    expect(onEnter).toHaveBeenCalledTimes(1);
    expect(onEntering).toHaveBeenCalledTimes(1);
    expect(onEntered).toHaveBeenCalledTimes(1);
    expect(onExit).toHaveBeenCalledTimes(1);
    expect(onExiting).toHaveBeenCalledTimes(1);
    expect(onExited).not.toHaveBeenCalled();

    act(() => {
      jest.runAllTimers();
    });
    expect(onEnter).toHaveBeenCalledTimes(1);
    expect(onEntering).toHaveBeenCalledTimes(1);
    expect(onEntered).toHaveBeenCalledTimes(1);
    expect(onExit).toHaveBeenCalledTimes(1);
    expect(onExiting).toHaveBeenCalledTimes(1);
    expect(onExited).toHaveBeenCalledTimes(1);
  });
});
