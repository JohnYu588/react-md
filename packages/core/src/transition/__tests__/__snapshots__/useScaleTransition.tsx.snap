// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScaleTransition should default using the horizontal transition 1`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
</div>
`;

exports[`ScaleTransition should default using the horizontal transition 2`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
  <div
    class="rmd-scale-transition--enter rmd-scale-transition--enter-active"
    data-testid="element"
  >
    This is some content.
  </div>
</div>
`;

exports[`ScaleTransition should default using the horizontal transition 3`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
  <div
    data-testid="element"
  >
    This is some content.
  </div>
</div>
`;

exports[`ScaleTransition should default using the horizontal transition 4`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
  <div
    class="rmd-scale-transition--exit rmd-scale-transition--exit-active"
    data-testid="element"
  >
    This is some content.
  </div>
</div>
`;

exports[`ScaleTransition should default using the horizontal transition 5`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
</div>
`;

exports[`ScaleTransition should merge the className and allow to be rendered by default 1`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
  <div
    class="custom-class"
    data-testid="element"
  >
    This is some content.
  </div>
</div>
`;

exports[`ScaleTransition should merge the className and allow to be rendered by default 2`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
  <div
    class="custom-class rmd-scale-transition--exit rmd-scale-transition--exit-active"
    data-testid="element"
  >
    This is some content.
  </div>
</div>
`;

exports[`ScaleTransition should merge the className and allow to be rendered by default 3`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
</div>
`;

exports[`ScaleTransition should merge the className and allow to be rendered by default 4`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
  <div
    class="custom-class rmd-scale-transition--enter rmd-scale-transition--enter-active"
    data-testid="element"
  >
    This is some content.
  </div>
</div>
`;

exports[`ScaleTransition should merge the className and allow to be rendered by default 5`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
  <div
    class="custom-class"
    data-testid="element"
  >
    This is some content.
  </div>
</div>
`;

exports[`ScaleTransition should render correctly for a vertical transition 1`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
</div>
`;

exports[`ScaleTransition should render correctly for a vertical transition 2`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
  <div
    class="rmd-scale-transition--enter rmd-scale-transition--y-enter rmd-scale-transition--enter-active rmd-scale-transition--y-enter-active"
    data-testid="element"
  >
    This is some content.
  </div>
</div>
`;

exports[`ScaleTransition should render correctly for a vertical transition 3`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
  <div
    data-testid="element"
  >
    This is some content.
  </div>
</div>
`;

exports[`ScaleTransition should render correctly for a vertical transition 4`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
  <div
    class="rmd-scale-transition--exit rmd-scale-transition--y-exit rmd-scale-transition--exit-active rmd-scale-transition--y-exit-active"
    data-testid="element"
  >
    This is some content.
  </div>
</div>
`;

exports[`ScaleTransition should render correctly for a vertical transition 5`] = `
<div>
  <button
    type="button"
  >
    Toggle
  </button>
</div>
`;
